bluepyopt/ephys/simulators.py

```diff
@@ -175,3 +179,90 @@ def __init__(self, message, original):
 
         super(NrnSimulatorException, self).__init__(message)
         self.original = original
+
+
+class ArbSimulator(object):
+
+    """Arbor simulator"""
+
+    def __init__(self, dt=None, ext_catalogues=None):
+        """Constructor
+
+        Args:
+            dt (float): the integration time step used by Arbor.
+            ext_catalogues (): Name to path mapping of non-Arbor built-in
+            NMODL mechanism catalogues compiled with modcc
+        """
+
+        self.dt = dt
+        self.ext_catalogues = ext_catalogues
+        if ext_catalogues is not None:
+            for cat, cat_path in ext_catalogues.items():
+                cat_lib = '%s-catalogue.so' % cat
+                cat_path = pathlib.Path(cat_path).resolve()
+                if not os.path.exists(cat_path / cat_lib):
+                    raise ArbSimulatorException(
+                        'Cannot find %s at %s - first build' % (cat_lib,
+                                                                cat_path) +
+                        ' mechanism catalogue with modcc:' +
+                        ' arbor-build-catalogue %s %s' % (cat, cat_path))
+        # TODO: add parameters for discretization
```

Comment:
Please turn this TODO into an issue and explain the following:

* state the importance of this TODO item
* will the code fail if this is left unimplemented?
* will the code assume a default behaviour if this is left unimplemented?

Reply:
The current code runs fine with defaults (cf. the examples and validation with Neuron). This adds more flexibility for the user to define control volumes (e.g. described in a [tutorial](https://docs.arbor-sim.org/en/latest/tutorial/single_cell_detailed.html#the-decorations)) that can be added by the simulator to the decor during execution.

Reply:
Thanks. Just to better understand: if users want to define control volumes today, they will provide the input accordingly but bluepyopt will not respond to it, right?

If that is the case, I would like a warning to be raised to deliver a message like this: "this version of bluepyopt does not support control volume definitions". Otherwise the users may be mistaken as they expect to see their input to be effective in this version of bluepyopt.

Reply:
One speculative comment concerning other possible Arbor features that are not yet supported on bluepyopt.

In general if there are some other features of Arbor that users may provide as part of their input to bluepyopt and if bluepyopt does not use them and goes silent about them it can be misleading.

If that is the case for any of the arbor features, bluepyopt should say: "this feature is not supported and will be ignored".

Reply:
If users want to define their own custom control volume policy, they would export a BluePyOpt model to JSON/ACC files and then edit the decor (adding smth as in the referenced tutorial), and then use that to run a simulation in Arbor as demonstrated in the [new tutorials](https://docs.arbor-sim.org/en/latest/tutorial/single_cell_bluepyopt.html). Currently, there's no way for a user to edit the decor generated in a BluePyOpt optimization with Arbor as this is generated on the fly in the `ArbSweepProtocol.run()`. Hence, the above hint in `simulators.py`.

To the second point - sure, the risk of unnoticed Arbor features silently failing is not so big, though, because the core idea of the Arbor support was not to extend the BluePyOpt frontend (that is unified between the simulators), but export what the user constructs to Arbor under the hood (in `ArbSweepProtocol`). So none of the typical Arbor objects are directly user-accessible.

Reply:
Ok sounds good